Paul Kilgarriff - Weekend 2 Challenge

PKilgarriff opened the pull request:
#  Paul Kilgarriff

Generally happy with the individual classes, but would like to develop a more sensible approach to having classes message one another. Mocking has been easier than before, but with more reliance on an external API than before I would like to know the best practice.

# User stories 

Please list which user stories you've implemented (delete the ones that don't apply).

- [x] User story 1: "I would like to see a list of dishes with prices"
- [x] User story 2: "I would like to be able to select some number of several available dishes"
- [x] User story 3: "I would like to check that the total I have been given matches the sum of the various dishes in my order"
- [x] User story 4: "I would like to receive a text such as "Thank you! Your order was placed and will be delivered before 18:52" after I have ordered"

# README checklist

Does your README contains instructions for

- [ ] how to install,
- [x] how to run,
- [x] and how to test your code?

[Here is a pill](https://github.com/makersacademy/course/blob/main/pills/readmes.md) that can help you write a great README!

PKilgarriff committed:
Drive creation of Menu, Dish , Takeaway and TwilioManager classes

 - test for adding a dish to a menu and listing dishes
   - add methods to menu:
     - #add_dish
     - list_dishes
   - add available? to Dish

PKilgarriff committed:
Update Dish class

- add set_unavailable method
- add set_available method

PKilgarriff committed:
Update Menu class

- add list_available_dishes method
- add tests for listing multiple dishes and only available dishes

PKilgarriff committed:
Create Order class

- add @current_items with attr_reader
- add add_to_order method
- add remove_from_order method
- add list_order method
- add order_total_price method
- add confirm_order method

PKilgarriff committed:
Add credentials file to .gitignore

PKilgarriff committed:
Add TwilioManager class

- require twilio-ruby gem and local credentials file
- add creation of Client in initialize
- add client_exists? method
- add send_sms: uses Twilio to send text to specific phone by default

PKilgarriff committed:
Add Takeaway class

- initialize with a TwilioManager instance
- add send_confirmation_text method
- add confirmation_string private method
- add time_in_an_hour private method
- test the public methods with Timecop and a twilio spy
- Add timecop to Gemfile

PKilgarriff committed:
Update Menu class

- add find_by_name method
- remove attr_reader for @dishes
- remove unused indexes from list methods

PKilgarriff committed:
Update Order class

- add select_items method
- test for select_items with a valid item on the menu

PKilgarriff committed:
Update Takeaway class

Pass in TwilioManager, Order and Menu as parameters for initialize
- Add interactive_menu
- print_options
- process
- quit_program
- current_order
Mock stdin for testing interactive menu

PKilgarriff committed:
Update Takeaway spec

Comment out test that runs interactive menu as not properly mocked
Currently calls 'exit' on rspec or gets stuck in infinite loop

PKilgarriff committed:
Swap PK notes with README for Code Review

jonnyabrams commented on README.md:
```diff
@@ -47,37 +41,154 @@ So that I am reassured that my order will be delivered on time
+  - Order
+  - Menu
+  - Dish
+  - TwilioManager
```
Hadn't occurred to me that Takeaway and Dish could/should have their own classes - props on the diagramming

jonnyabrams replied:
Hadn't occurred to me that Takeaway and Dish could/should have their own classes - props on the diagramming

jonnyabrams commented on README.md:
````diff
@@ -47,37 +41,154 @@ So that I am reassured that my order will be delivered on time
+order = Order.new
+takeaway = Takeaway.new(twilio, menu, order)
+takeaway.interactive_menu
+```
````
Very detailed instructions for use, I must emulate this

PKilgarriff replied:
Thank you, it was also very helpful for me running it quickly in irb without having to manually type multiple lines of code!

jonnyabrams commented on lib/order.rb:
```diff
@@ -0,0 +1,40 @@
+  def select_items(menu)
+    loop do
+      puts 'Choose a dish to add to your order'
+      selection = gets.chomp
```
Ah, user input, my old friend

PKilgarriff replied:
It's not fully working yet - fine in irb, but the mocking of stdin is an absolute nightmare (or at least I've made it one)

jonnyabrams commented on lib/takeaway.rb:
```diff
@@ -0,0 +1,65 @@
+  def quit_program
+    puts "Exiting program..."
+    exit # Needs to be commented out for running tests
+  end
```
This is incredibly thorough, nice one

PKilgarriff committed:
Update Menu class

- Removed commented attr_reader for dishes
- add explanatory comment for add_dish
